code/go/internal/validator/semantic/kibana_matching_object_ids.go

```diff
@@ -0,0 +1,55 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package semantic
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	ve "github.com/elastic/package-spec/code/go/internal/errors"
+
+	"github.com/elastic/package-spec/code/go/internal/pkgpath"
+	"github.com/pkg/errors"
+)
+
+// ValidateKibanaObjectIDs returns validation errors if there are any Kibana
+// object files that define IDs not matching the file's name. That is, it returns
+// validation errors if a Kibana object file, foo.json, in the package defines
+// an object ID other than foo inside it.
+func ValidateKibanaObjectIDs(pkgRoot string) ve.ValidationErrors {
+	var errs ve.ValidationErrors
+
+	filePaths := filepath.Join(pkgRoot, "kibana", "*", "*.json")
+	objectFiles, err := pkgpath.Files(filePaths)
+	if err != nil {
+		errs = append(errs, errors.Wrap(err, "error finding Kibana object files"))
+		return errs
+	}
+
+	for _, objectFile := range objectFiles {
+		filePath := objectFile.Path()
+
+		idPath := "$.id"
+		// Special case: object is of type 'security_rule'
+		if filepath.Base(filepath.Dir(filePath)) == "security_rule" {
+			idPath = "$.rule_id"
+		}
+
+		objectID, err := objectFile.Values(idPath)
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "unable to get Kibana object ID in file [%s]", filePath))
+			continue
+		}
+
+		fileID := strings.Replace(filepath.Base(filePath), filepath.Ext(filePath), "", -1)
+		if fileID != objectID {
```

Comment:
oh, it doesn't complain about different type here? (string vs interface)

Reply:
Yes, this works as explained in https://golang.org/ref/spec#Comparison_operators:

> Interface values are comparable. Two interface values are equal if they have identical dynamic types and equal dynamic values or if both have value `nil`.
> A value `x` of non-interface type `X` and a value `t` of interface type `T` are comparable when values of type `X` are comparable and `X` implements `T`. They are equal if `t`'s dynamic type is identical to `X` and `t`'s dynamic value is equal to `x`.

Reply:
TIL!